docs/02-app/01-building-your-application/02-data-fetching/03-forms-and-mutations.mdx

```diff
@@ -391,9 +393,40 @@ function SubmitButton() {
 }
 ```
 
-> **Good to know:**
->
-> - Displaying loading or error states currently requires using Client Components. We are exploring options for server-side functions to retrieve these values as we move forward in stability for Server Actions.
```

Comment:
Would love some clarification here!

> Displaying loading or error states currently requires using Client Components.

Isn't this still true?

https://github.com/vercel/next.js/pull/55417#issuecomment-1722534196

From what I've understood, this is a client-only component that _also_ supports progressive enhancement. This means that it now works with SSR, but not server components.

Reply:
Correct, using these hooks (`useFormState` and `useFormStatus`) require using Client Components. [Client Components are prerendered](https://github.com/reactwg/server-components/discussions/4), so they work with dynamic rendering.

Reply:
Wonderful, thank you. I wonder if it's worth clarifying that there still isn't a way to get the result of an action in a server-only form / server component?